libs/auth/react/src/lib/auth/AuthProvider.tsx

```diff
@@ -0,0 +1,260 @@
+   */
+  basePath: string
+  children: ReactNode
+}
+
+const getReturnUrl = (returnUrl: string, { redirectPath }: AuthSettings) => {
+  if (redirectPath && returnUrl.startsWith(redirectPath)) {
+    return '/'
+  }
+
+  return returnUrl
+}
+
+export const AuthProvider = ({
+  children,
+  autoLogin = true,
+  basePath,
+}: AuthProviderProps) => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+  const [hasError, setHasError] = useState(false)
+  const userManager = getUserManager()
+  const authSettings = getAuthSettings()
+  const monitorUserSession = !authSettings.scope?.includes('offline_access')
+  const urlRef = useRef<string>()
+  urlRef.current = `${window.location.pathname}${window.location.search}`
+
+  const signIn = useCallback(async () => {
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+
+    return userManager.signinRedirect({
+      state: getReturnUrl(urlRef.current as string, authSettings),
+    })
+    // Nothing more happens here since browser will redirect to IDS.
+  }, [dispatch, userManager, authSettings, urlRef])
+
+  const signInSilent = useCallback(async () => {
+    let user = null
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+    try {
+      user = await userManager.signinSilent()
+      dispatch({ type: ActionType.SIGNIN_SUCCESS, payload: user })
+    } catch (error) {
+      console.error('AuthProvider: Silent signin failed', error)
+      dispatch({ type: ActionType.SIGNIN_FAILURE })
+    }
+
+    return user
+  }, [userManager, dispatch])
+
+  const switchUser = useCallback(
+    async function switchUser(nationalId?: string, newLocation?: Location) {
+      if (newLocation) {
+        urlRef.current = `${newLocation.pathname}${newLocation.search}`
+      }
+
+      const args =
+        nationalId !== undefined
+          ? {
+              login_hint: nationalId,
+              /**
+               * TODO: remove this.
+               * It is currently required to switch delegations, but we'd like
+               * the IDS to handle login_required and other potential road
+               * blocks. Now OidcSignIn is handling login_required.
+               */
+              prompt: 'none',
+            }
+          : {
+              prompt: 'select_account',
+            }
+
+      dispatch({
+        type: ActionType.SWITCH_USER,
+      })
+      return userManager.signinRedirect({
+        state:
+          authSettings.switchUserRedirectUrl ??
+          getReturnUrl(urlRef.current as string, authSettings),
+        ...args,
+      })
+      // Nothing more happens here since browser will redirect to IDS.
+    },
+    [userManager, dispatch, authSettings, urlRef],
+  )
+
+  const signOut = useCallback(
+    async function signOut() {
+      dispatch({
+        type: ActionType.LOGGING_OUT,
+      })
+      await userManager.signoutRedirect()
+    },
+    [userManager, dispatch],
+  )
+
+  const checkLogin = useCallback(async () => {
```

Comment:
Fyi, using naming functions inside hooks (and any place they're not assigned directly to a variable) can help when debugging, eg the names show up in stack traces.